### DISCUSSION TITLE
set_edge_attributes_from_iter

### DISCUSSION THREAD

**Discussion by amorfinv** [networkx/networkx#5322] [Ideas]
Hello,

I was wondering if it would be useful to extend ```networkx.classes.function.set_edge_attributes()``` or make a new function called ```networkx.classes.function.set_edge_attributes_from_iter()```

```python
def set_edge_attributes_from_iter(G, data_array, attr_name):
    """
    Set the attribute of the edges of the graph G from an iterable.
    """
    for i, (u, v, data) in enumerate(G.edges(data=True)):
        data[attr_name] = data_array[i]
```

That way we don't always have to feed the attribute data as a dictionary. 

I am doing some simulations where the attributes dynamically change and created this function that is almost the same as ```networkx.classes.function.set_edge_attributes()```.

If useful, I can create a pull request.

Best,
Andres

**Comment by rossbar** [comment]
In principle this is possible now that the edges are guaranteed to be ordered, but my gut reaction is that adding API for this would be a bad idea. The graph edges are mutable, so if a user modifies the graph (removes an edge, say) and doesn't "pop" the appropriate field from the data_array, then you will silently get the wrong answer the next time you call `set_edge_attributes_from_iter`.

If you are in a situation where you *know* that the graph is not changing, just the attributes of it's components, then the above is fine - but that assumption doesn't apply to the set of use-cases for networkx graphs.

**Reply by amorfinv** [reply]
Thanks for the reply! In my current use case I would not think of changing the edges mid-simulation. I do see how that could be a problem. 

Perhaps there could be an option to feed (1) a list of edges (2) the ```data_array```,  (3) the attribute name. If a list of edges isn't given then it assumes the number of edges is equal to length of ```data_array```. 

Although, at that point it is a bit closer to the original function. However, it might still be preferable than building dictionaries, especially if it's a large graph and we can feed in numpy arrays.

**Comment by dschult** [comment]
It's pretty straightforward to construct the dict. 
```python
nx.set_edge_attributes(G, dict(zip(G.edges, data_array)), attr)
```
And if you are coming from a numpy array you should maybe look at `nx.from_pandas_edgelist` or maybe better `nx.from_numpy_array` though that is an adjacency matrix instead of an edgelist.

**Reply by amorfinv** [reply]
That is a good suggestion thank you... I guess from my current case it is not necessary to pass the edges since they do not change and I am changing the attributes of many edges at once. 

It was just a bit redundant pass the edge information separately since it is already in the graph.
